README
fancypants: a library of data visualisation trinkets

Currently there are javascript and python libraries. More may follow.

The software is free, in all senses of the word. For further
details see UNLICENSE

python/CHANGES
1.3   	Added support for nested data structures
		Made everything a bit more OO
		Added examples

1.4		Renamed Area class to Rect

python/tests.py
from fancypants import Rect, Dataset, Point
import unittest

class TestDataset(unittest.TestCase):
    """Unit tests for the Dataset class"""

    def setUp(self):
        """docstring for setUp"""
        pass
    
    def tearDown(self):
        """docstring for tearDown"""
        pass
    
    def testTreeMap(self):
        data = [('a', 100), ('b', 150), ('c', 250)]
        area = Rect(300,200)

        # will generate a treemap like this:
        # 
        # (0,0)+------------+------------+
        #      |   c        |  b         |
        #      | (150x200)  | (150x120)  |
        #      |            |            |
        #      |            +------------+
        #      |            |  a         |
        #      |            | (150x80)   |
        #      +------------+------------+
        
        ds = Dataset(data)
        tm = ds.treemap(area)
        
        expected = [
          # [x,   y,   w,   h,   val, lbl]
            [0,   0,   150, 200, 250, 'c'],
            [150, 0,   150, 120, 150, 'b'],
            [150, 120, 150, 80,  100, 'a']
        ]
        
        for i in range(0, len(expected)):
            self.assertEqual(tm[i].origin.x,    expected[i][0])
            self.assertEqual(tm[i].origin.y,    expected[i][1])
            self.assertEqual(tm[i].area.width,  expected[i][2])
            self.assertEqual(tm[i].area.height, expected[i][3])
            self.assertEqual(tm[i].value,       expected[i][4])
            self.assertEqual(tm[i].label,       expected[i][5])
                                
            # test properties
            self.assertEqual(tm[i].x,           expected[i][0])
            self.assertEqual(tm[i].y,           expected[i][1])
            self.assertEqual(tm[i].width,       expected[i][2])
            self.assertEqual(tm[i].height,      expected[i][3])

    def testNestedTreeMap(self):
        data = [
            ('a', [('p', 10),  ('q', 40)]),
            ('b', [('r', 150), ('s', 50)])
        ]
        area = Rect(250,200)

        # will generate a treemap like this:
        # 
        # (0,0)+--------------+------------+
        #      |              |            |
        #      |   b.r        |  a.q       |
        #      | (200x150)    | (50x160)   |
        #      |              |            |
        #      +--------------+            |
        #      |              +------------+
        #      |   b.s        | a.p        |
        #      | (200x50)     | (50x40)    |
        #      +--------------+------------+

        ds = Dataset(data)
        tm = ds.treemap(area)
        
        expected = [
          # [x,   y,   w,   h,   val,  lbl]
            [0,   0,   200, 150, 150, 'b.r'],
            [0,   150, 200, 50,  50,  'b.s'],
            [200, 0,   50,  160, 40,  'a.q'],
            [200, 160, 50,  40,  10,  'a.p'],
        ]

        for i in range(0, len(expected)):
            self.assertEqual(tm[i].origin.x,    expected[i][0])
            self.assertEqual(tm[i].origin.y,    expected[i][1])
            self.assertEqual(tm[i].area.width,  expected[i][2])
            self.assertEqual(tm[i].area.height, expected[i][3])
            self.assertEqual(tm[i].value,       expected[i][4])
            self.assertEqual(tm[i].label,       expected[i][5])

            # test properties
            self.assertEqual(tm[i].x,           expected[i][0])
            self.assertEqual(tm[i].y,           expected[i][1])
            self.assertEqual(tm[i].width,       expected[i][2])
            self.assertEqual(tm[i].height,      expected[i][3])

if __name__ == '__main__':
    unittest.main()

javascript/fancypants.js
function Rect(width, height) {
    this.width = width;
    this.height = height;
}

function Point(x, y) {
    this.x = x;
    this.y = y;
}

function Frame (label, value, origin, area) {
    this.label = label;
    this.value = value;
    this.origin = origin;
    this.area = area;
}

function TreeMap() {
    this.frames = [];
}

TreeMap.prototype.push = function(data) {
    this.frames.push(data);
}

TreeMap.prototype.concat = function(data) {
    if (data instanceof TreeMap) {
        this.frames = this.frames.concat(data.frames);
    } else {
        throw "Argument to TreeMap.concat must be a TreeMap object";
    }
}

function BaseData() { }

BaseData.prototype.get_total = function() {
    throw "Not implemented";
};

BaseData.prototype.get_label = function() {
    return this.label;
};

function DataPoint(value, label) {
    if (value < 0) {
        throw "value arg to DataPoint constructor must be >= 0";
    }
    this.value = value;
    this.label = label;
}

// DataPoint inherits from BaseData
// see https://developer.mozilla.org/en/Introduction_to_Object-Oriented_JavaScript
DataPoint.prototype = new BaseData();
DataPoint.prototype.constructor = DataPoint;

DataPoint.prototype.get_total = function() {
    return this.value;
};

DataPoint.prototype.get_value = function() {
    return this.value;
};

DataPoint.prototype.get_label = function() {
    return this.label;
};

function Dataset(data, label) {
    this.data = [];
    this.label = label;
    this.total = null;
    
    for (var i = 0; i < data.length; i++) {
        var datum = data[i];
        var label = datum[0];
        var value = datum[1];
        
        if (this.label && label) {
            label = this.label + '.' + label;
        }
        
        if (value instanceof Array) {
            this._append_data(new Dataset(value, label));
        } else {
            this._append_data(new DataPoint(value, label));
        }
    }
    this.sort();
}

Dataset.prototype = new BaseData();
Dataset.prototype.constructor = Dataset;

Dataset.prototype.sort = function() {
    this.data.sort(
        function(a, b) {
            var a_total = a.get_total();
            var b_total = b.get_total();
            if (a_total < b_total) { return 1 };
            if (a_total > b_total) { return -1 };
            return 0;
        }
    )
};

Dataset.prototype._append_data = function(data) {
    this.data.push(data)
};

Dataset.prototype.append_data = function(data) {
    this._append_data(data);
    this.sort();
};

Dataset.prototype.get_total = function() {
    if (this.total == null) {
        total = 0;
        for (var i = 0; i < this.data.length; i++) {
            total = total + this.data[i].get_total();
        }
        this.total = total;
    }
    return this.total;
};

Dataset.prototype.treemap = function(area, origin, padding, threshold, flat) {
    origin      = origin || new Point(0,0);
    padding     = padding || 0;
    threshold   = threshold || 0;
    if (flat == null)
        flat = true;
    
    var total = this.get_total();
    var x = origin.x;
    var y = origin.y;
    var w = area.width;
    var h = area.height;
    
    var result = new TreeMap();
    
    for (var i = 0; i < this.data.length; i++) {
        var datum = this.data[i];
        var is_last = false;
        
        if (total < threshold) {
            var label_tokens = datum.label.split('.');
            label_tokens[-1] = "Others";
            label = label_tokens.join('.');
            datum = new DataPoint(total, label);
            is_last = true;
        }
        
        if (i == this.data.length - 1) {
            is_last = true;
        }
        
        var this_x = x;
        var this_y = y;
        var this_w = w;
        var this_h = h;
        
        var fixed_height = w > h;
        
        var proportion = datum.get_total() / total;
        total -= datum.get_total();
        
        if (fixed_height) {
            this_w  = Math.floor(w * proportion);
            w       = w - this_w;
            x       = x + this_w;
        } else {
            this_h  = Math.floor(h * proportion);
            h       = h - this_h;
            y       = y + this_h;
        }
        
        // add padding
        if (fixed_height && !is_last) {
            this_w -= padding;
        } else if (!fixed_height && !is_last) {
            this_h -= padding;
        }
        
        // sanity check: did we end up rendering a negative-sized frame?
        if (this_w < 0) {
            this_w = 0;
        }
        if (this_h < 0) {
            this_h = 0;
        }
        
        origin = new Point(this_x, this_y);
        area   = new Rect(this_w, this_h);
        
        if (datum instanceof DataPoint) {
            result.push(new Frame(datum.label, datum.value, origin, area));
        } else {
            var tmap = datum.treemap(area, origin, padding, threshold, flat);
            if (flat) {
                result.concat(tmap);
            } else {
                result.push(tmap);
            }
        }
    }
    
    return result;
}

Dataset.prototype.treemap_in_element = function(element_id, padding, threshold) {
    var element = document.getElementById(element_id);
    var w       = parseInt(window.getComputedStyle(element, null).getPropertyValue('width'));
    var h       = parseInt(window.getComputedStyle(element, null).getPropertyValue('height'));
    var area    = new Rect(w, h);
    var origin  = new Point(0, 0);
    
    if (padding == null) {
        padding = 0;
    }
    if (threshold == null) {
        threshold = 0;
    }
    
    var tm      = this.treemap(area, origin, padding, threshold, true);
    
    for (var i = 0; i < tm.frames.length; i++) {
      var frame = tm.frames[i];
      var div = document.createElement('div');
      div.style.position = "absolute";
      div.className     = "frame " + frame.label.replace('.', ' ');
      div.style.width   = frame.area.width;
      div.style.height  = frame.area.height;
      div.style.left    = frame.origin.x;
      div.style.top     = frame.origin.y;
      element.appendChild(div);
      
      var content = document.createElement('div');
      content.className = 'content';
      div.appendChild(content);

      var val = document.createElement('div');
      val.className = 'value';
      val.innerHTML = frame.value;
      content.appendChild(val);

      var lbl = document.createElement('div');
      lbl.className = 'label';
      lbl.innerHTML = frame.label;
      content.appendChild(lbl);
    }
}


python/examples/html/oecd-broadband.py
from fancypants import Dataset, Rect, Point

# Example: OECD data on broadband subscriber statistics
# http://www.guardian.co.uk/news/datablog/2009/may/20/broadband-internetphonesbroadband

data = [
    ('Australasia', [
        ('Australia', 5368000),
        ('New Zealand', 914961),
    ]),
    ('Europe', [
        ('Austria', 1792408),
        ('Belgium', 2962450),
        ('Czech Republic', 1769684),
        ('Denmark', 2021404),
        ('Finland', 1616900),
        ('France', 17725000),
        ('Germany', 22532000),
        ('Greece', 1506614),
        ('Hungary', 1696714),
        ('Iceland', 99883),
        ('Ireland', 896346),
        ('Italy', 11283000),
        ('Luxembourg', 141584),
        ('Netherlands', 5855000),
        ('Norway', 1607750),
        ('Poland', 3995458),
        ('Portugal', 1692306),
        ('Slovak Republic', 618871),
        ('Spain', 9156969),
        ('Sweden', 2905000),
        ('Switzerland', 2533643),
        ('United Kingdom', 17275660),
    ]),
    ('North America', [
        ('Canada', 9577648),
        ('United States', 80071074),
    ]),
    ('Asia', [
        ('Japan', 30107327),
        ('Korea', 15474931),
        ('Turkey', 5736619),
    ]),
    ('South America', [
        ('Mexico', 7604629),
    ]),
]

ig = Dataset(data)

frames = ig.treemap(Rect(600,420), origin=Point(0,0), padding=1, threshold=40000000)

with open("oecd-broadband.html", "w") as f:

    f.write("""<html>
<head>
    <title>fancypants example: OECD broadband subscriptions</title>
    <style>
        @import url('oecd-broadband.css');
    </style>
</head>
<body>
<p>Broadband subscribers (millions) per OECD country. Data source: <a href="http://www.guardian.co.uk/news/datablog/2009/may/20/broadband-internetphonesbroadband">Guardian Datablog</a>. Created using <a href="http://github.com/simonwhitaker/fancypants">fancypants</a></p>
<div style="position:relative">
""")

    for frame in frames:
        f.write("""<div class='frame %s' style='position: absolute; top: %ipx; left: %ipx; width: %ipx; height: %ipx'>
    <div class="text">
        <div class="value">%i</div>
        <div class="label">%s</div>
    </div>
</div>
""" %
            (
                frame.label.replace(' ', '_').replace('.', ' '),
                frame.y,
                frame.x,
                frame.width,
                frame.height,
                round(frame.value/1000000),
                frame.label.split('.')[-1],
            )
        )

    f.write("</div></body></html>")

python/fancypants.py
class Rect(object):
    def __init__(self, width, height):
        super(Rect, self).__init__()
        self.width = width
        self.height = height

class Point(object):
    def __init__(self, x, y):
        super(Point, self).__init__()
        self.x = x
        self.y = y

class Frame(object):
    def __init__(self, label, value, origin, area):
        super(Frame, self).__init__()
        self.label = label
        self.value = value
        self.origin = origin
        self.area = area
    
    @property
    def width(self):
        return self.area.width
    
    @property
    def height(self):
        return self.area.height
    
    @property
    def x(self):
        return self.origin.x
    
    @property
    def y(self):
        return self.origin.y
    
    def __str__(self):
        return "%s (%s): (%s, %s), (%s, %s)" % (self.value, self.label, self.origin.x, self.origin.y, self.area.width, self.area.height)

class BaseData(object):
    """Base class for data objects"""
    def __init__(self):
        super(BaseData, self).__init__()
        self.label = None
        self.value = None
    
    def get_total(self):
        """Returns total value of the data object"""
        raise NotImplementedError()
    
    def get_label(self):
        return self.label

class DataPoint(BaseData):
    """docstring for DataPoint"""
    def __init__(self, value, label=None):
        super(DataPoint, self).__init__()
        
        if value < 0:
            raise ValueError("All data must contain positive values (%.2f, %s)" % (value, label))
        
        self.label = label
        self.value = value
    
    def get_total(self):
        return self.value
    
    def get_value(self):
        return self.value

class Dataset(object):
    def __init__(self, data, label=None):
        super(Dataset, self).__init__()
        self.data = []
        self.label = label
        self.total = None
        
        # print "Dataset ctor called with %s, %s" % (data, label)
        
        for datum in data:
            (label, value) = datum
            if self.label and label:
                label = "%s.%s" % (self.label, label)
            
            if (isinstance(value, list)):
                self._append_data(Dataset(value, label))
            else:
                self._append_data(DataPoint(value, label))
        
        # Sort the data, largest to smallest
        self.sort()
    
    def sort(self):
        # Sort the data, largest to smallest
        self.data.sort(cmp=lambda a,b: cmp(b.get_total(), a.get_total()))
    
    def _append_data(self, data):
        self.data.append(data)
        # Unset total so it gets recalculated on next call to get_total
        self.total = None
    
    def append_data(self, data):
        self._append_data(data)
        self.sort()
    
    def get_total(self):
        if self.total is None:
            total = 0
            for datum in self.data:
                total += datum.get_total()
            self.total = total
        return self.total
    
    def treemap(self, area, origin=Point(0,0), padding=0, threshold=0, flat=True):
        """
        treemap: returns a set of Frame objects that describe a treemap
        
        Arguments:
            area:      A Rect object, describing the size of the treemap
            origin:    A Point object, describing the origin of the treemap in 2D space
            padding:   padding between the frames in the treemap (defaults to 0)
            threshold: if total < threshold, aggregate all remaining values into an "other" frame 
            flat:      if nested data is provided, provide a flat array as output, rather
                       than nested arrays
    
        Return value is a list of Frames
        """
    
        # TODO:
        # Enforce some sane threshold? Otherwise can get lots of tiny frames
        if len(self.data) == 0:
            raise Exception("Your infographic needs at least one data point!")
    
        total = self.get_total()
    
        # Coords for all of data
        x, y = origin.x, origin.y
        w, h = area.width, area.height
        
        result = []
    
        for i in range(0, len(self.data)):
            try:
                datum = self.data[i]
            except IndexError:
                print "index %i out of range for data of length %i" % (i, len(data))
                raise IndexError
                
            is_last = False
            
            # If there's a threshold and what's left sums to less than the threshold,
            # aggregate all the remaining data under "Others"
            # print "checking threshold, comparing %i < %i" % (total, threshold)
            if total < threshold:
                label_tokens = datum.label.split('.')
                label_tokens[-1] = 'Others'
                label = '.'.join(label_tokens)
                datum = DataPoint(total, label)
                is_last = True
    
            if i == len(self.data) - 1:
                is_last = True
    
            # Coords for this panel
            this_x, this_y  = x, y
            this_w, this_h  = w, h
            
            # If w > h then we'll fix the height to h and set
            # w to a proportion, else do the reverse
            fixed_height = w > h
    
            proportion = float(datum.get_total()) / total
            total -= datum.get_total()
    
            if fixed_height:
                this_w  = int(float(w) * proportion)
                w       = w - this_w
                x       = x + this_w
    
            else:
                this_h  = int(float(h) * proportion)
                h       = h - this_h
                y       = y + this_h
    
            # Adding the padding...
            # If we're working on a fixed height, reduce the height by another 1 * padding
            # If we're working on a fixed width, reduce the width likewise
            # Special case: if this is the last frame, reduce neither
            if fixed_height is True and not is_last:
                this_w -= padding
            
            if fixed_height is False and not is_last:
                this_h -= padding
            
            # Sanity check: did we end up rendering a negative-sized frame?
            if this_w < 0:
                this_w = 0
            if this_h < 0:
                this_h = 0
            
            origin = Point(this_x, this_y)
            area   = Rect(this_w, this_h)
            
            if (isinstance(datum, DataPoint)):
                result.append(Frame(label=datum.label, value=datum.value, origin=origin, area=area))
            else:
                tmap = datum.treemap(area, origin, padding=padding, flat=flat, threshold=threshold)
                if flat:
                    result = result + tmap
                else:
                    result.append(tmap)
            if is_last:
                break
        
        return result


if __name__ == '__main__':
    ig = Dataset(
        [
            ('foo', 50), 
            ('bar', 50),
            ('a', [
                ('foo', 20), 
                ('bar', 30),
                ('wibble', [
                    ('oof', 40),
                    ('boof', 50)
                ])
            ])
        ]
    )
    frames = ig.treemap(Rect(300,200), padding=0)
    for f in frames:
        print f

    print "-------------------------"

    ig.append_data(DataPoint(110, 'blah'))
    frames = ig.treemap(Rect(300,200), padding=0)
    for f in frames:
        print f
